Adding filter of summarized field grouped by latitude and longitude on a map changes visualization type

kamilmielnik opened the pull request:
Fixes #32075

Slack discussions for context:
- https://metaboat.slack.com/archives/C505ZNNH4/p1686916009184259?thread_ts=1686832713.430189&cid=C505ZNNH4
- https://metaboat.slack.com/archives/C057T1QTB3L/p1687761720957009

# Description

This PR changes the behavior of the notebook editor in order to fix a bug.

## Before

Every time a (new or existing) question is edited on the fly and user clicks "visualize", the app will try to automatically pick the visualization type for the user (as long as the user hasn't hand-picked a visualization type for it earlier). In case none of the conditions in the code are met, the display is being reset to the default "table" mode. The [conditions in code don't cover all cases and don't account for nested queries](https://github.com/metabase/metabase/blob/9da0bd7/frontend/src/metabase-lib/Question.ts#L345-L427).

## After

This PR changes the above logic so that if we are about to reset the visualization type specifically to "table" - we won't and will keep the current visualization type as long as the current visualization type makes sense for the data.

In other words: once a question has been visualized, subsequent visualizations after editing the question will try to reuse the same visualization type, even when it has been automatically chosen.

## How to verify

### Scenario 1 - adding filter to a specific summarization

1. Create a question [like this](https://github.com/metabase/metabase/assets/6830683/f4cd18fd-a810-420e-aeb9-bcdfcee28dfc)
2. Visualize it (it's a map by default)
3. Go back to edit the question
4. Add a filter (e.g. count > 2)
5. Visualize it

Before: visualization type changes to table
After: visualization type stays a map (because it already was "map" in step 2 and it still makes sense to show it like that)

### Scenario 2 - generic summarization with more breakouts
1. Create a question [like this](https://github.com/metabase/metabase/assets/6830683/f4cd18fd-a810-420e-aeb9-bcdfcee28dfc) (the same as in scenario 1)
2. Visualize it (it's a map by default)
3. Go back to edit the question
4. Add "Birth Date" as another column to group by
5. Visualize it

Before: visualization type changes to table
After: visualization type stays a map (because it already was "map" in step 2 and it still makes sense to show it like that)

### Scenario 3 - generic summarization with more aggregations
1. Create a question [like this](https://github.com/metabase/metabase/assets/6830683/f4cd18fd-a810-420e-aeb9-bcdfcee28dfc) (the same as in scenario 1)
2. Visualize it (it's a map by default)
3. Go back to edit the question
4. Add "Average of Longitude" as another dimension to summarize by
5. Visualize it

Before: visualization type changes to table
After: visualization type stays a map (because it already was "map" in step 2 and it still makes sense to show it like that)

### Scenario 4 - removing a breakout column so that display does not make sense anymore

1. Create a question [like this](https://github.com/metabase/metabase/assets/6830683/f4cd18fd-a810-420e-aeb9-bcdfcee28dfc) (the same as in scenario 1)
2. Visualize it (it's a map by default)
3. Go back to edit the question
4. Remove "Latitude: Auto binned" breakout
5. Visualize it

Before: visualization type changes to bar chart, because map no longer makes sense
After: the same as before, nothing should change

kamilmielnik committed:
Take sourceQuery into account if StructuredQuery does not have any aggregations or breakouts when determining Question's default display
- It's an attempt to fix #30057

kamilmielnik committed:
Don't check source query for models

WiNloSt commented on frontend/src/metabase-lib/Question.ts:
```diff
@@ -353,6 +353,18 @@ class QuestionInner {
+        aggregations.length === 0 &&
+        breakouts.length === 0 &&
+        sourceQuery &&
+        !this.isDataset()
```
I'm not sure why this needs to check that the query isn't a model?

kamilmielnik replied:
> models always have to have display: "table"

Source: https://metaboat.slack.com/archives/C505ZNNH4/p1686833371108729?thread_ts=1686832713.430189&cid=C505ZNNH4

kamilmielnik replied:
I made the condition explicit in this commit: 4cc5669708807bfd748e98f144db2d086d8c206d

```tsx
      if (this.isDataset()) {
        return this.setDisplay("table");
      }
```

WiNloSt reviewed (commented):
With my limited knowledge of Metabase lib I'd say this fix looks alright. But would love feedback from Metabase-lib veteran still.

alxnddr reviewed (commented):
I think we should not consider source query aggregations and dimensions in all cases like that, because it leads to wrong results with different charts. This query is expected to be a table, but after the changes it will incorrectly assume the display type should be `line` based on the source query
<img width="1215" alt="Screen Shot 2023-06-16 at 4 45 21 PM" src="https://github.com/metabase/metabase/assets/14301985/9eb5b476-670e-41f2-a111-c62faa0eefea">

alxnddr commented:
Wdyt about using `sourceQuery` to determine the default display type but only if the original query contains filters only?

kamilmielnik committed:
Merge branch 'master' into 30057-aggregations-breakouts-fallback

kamilmielnik committed:
Use source query breakouts and aggregations to determine default display only if the question does not have any breakouts, aggregations, and joins
- Make the condition to always use "table" display for Models more prominent

kamilmielnik commented:
> Wdyt about using `sourceQuery` to determine the default display type but only if the original query contains filters/sorting only?

Sounds good @alxnddr :+1: I updated the code: 4cc5669708807bfd748e98f144db2d086d8c206d

This fix is very specific to these particular reproduction steps which makes me wonder - are there any other similar scenarios (i.e. questions) that could be affected here but we're missing?

ranquild reviewed (changes requested):
Discussed in Slack https://metaboat.slack.com/archives/C057T1QTB3L/p1687761720957009, currently investigating different approaches with locking/unlocking the display type.

kamilmielnik committed:
Merge branch 'master' into 30057-aggregations-breakouts-fallback

kamilmielnik committed:
Revert first attempt

kamilmielnik committed:
Introduce CardObject and fix displayIsLocked typing

kamilmielnik committed:
Refactor maybeUnlockDisplay into maybeResetDisplay

kamilmielnik committed:
Revert "Introduce CardObject and fix displayIsLocked typing"

This reverts commit 770e7a50d3b9f381741ba7cb1655c1d9be5f295c.

kamilmielnik committed:
Improve diff

kamilmielnik commented on frontend/src/metabase-lib/Question.ts:
```diff
@@ -311,16 +311,24 @@ class QuestionInner {
     const shouldUnlock = wasSensible && !isSensible;
-    const locked = this.displayIsLocked() && !shouldUnlock;
-    return this.setDisplayIsLocked(locked);
+    const defaultDisplay = this.setDefaultDisplay().display();
+
+    if (isSensible && defaultDisplay === "table") {
+      // any sensible display is better than the default table display
+      return this;
+    }
```
These lines make the difference vs the original code.

ranquild reviewed (commented):
Code wise looks good, lets run the test suite and see what fails

kamilmielnik committed:
Update unit tests after maybeUnlockDisplay to maybeResetDisplay refactoring

kamilmielnik committed:
Update maybeResetDisplay unit tests

kamilmielnik committed:
Improve test case names

kamilmielnik committed:
Add tag for the test that reproduces the issue

kamilmielnik committed:
Merge branch 'master' into 30057-aggregations-breakouts-fallback

kamilmielnik committed:
Add test for issue #30057
- use visualize() helper
- extract addCountGreaterThan2Filter()

kamilmielnik commented on frontend/test/metabase-lib/lib/Question.unit.spec.js:
```diff
@@ -459,34 +458,67 @@ describe("Question", () => {
 
         expect(question.displayIsLocked()).toBe(false);
+        expect(question.display()).not.toBe("funnel");
+        expect(sensibleDisplays).toContain(question.display());
```
Semantically, this is inversed - I don't expect array to contain the element - I expect the element to be in an array.

Is there anything similar to `expect(item).toBeIn(array)`?

WiNloSt replied:
Is there any reason why we don't just test if the display is a certain type i.e. why don't we
```ts
expect(question.display()).not.toBe("table");
```

kamilmielnik replied:
The same explanation applies here: https://github.com/metabase/metabase/pull/31646/files#r1250946190

kamilmielnik committed:
Add a test case for adding columns

kamilmielnik committed:
Add a test case for removing columns (resetting to default value)

kamilmielnik committed:
Add a test case for adding aggregations

kamilmielnik commented:
I'm re-requesting from anyone who already was involved in this PR and requesting review from
- all frontend teams because this change affects `metabase-lib`.
- @cdeweyx for [one final look](https://metaboat.slack.com/archives/C057T1QTB3L/p1687957682679209?thread_ts=1687761720.957009&cid=C057T1QTB3L) as this change needs approval from the product team

kamilmielnik commented:
> I just realized this still needs fixes for the metric (step 4 in reproduction steps). Apologies for premature review request.

I've extracted the issue this PR fixes out of #30057 as #32075.

----

I'm re-requesting from anyone who already was involved in this PR: @cdeweyx, @WiNloSt, @alxnddr .

WiNloSt commented on frontend/test/metabase-lib/lib/Question.unit.spec.js:
```diff
@@ -459,34 +458,67 @@ describe("Question", () => {
+        expect(question.display()).toBe("scalar");
+      });
+
+      it("should keep any sensible display when display was not locked (metabase#30057)", () => {
```
I'm not sure I understand this test case. This emphasizes the condition "when display was not locked". But the test would still pass when we call `.lockDisplay()` on the question.

I think this test exercises this code.
https://github.com/metabase/metabase/blob/f3583fae60d5281f9bcbaed18046fd8a4397ad66/frontend/src/metabase-lib/Question.ts#L322-L325

maybe the test name should be something along the line "should keep any sensible display when the default display is table"?

kamilmielnik replied:
> But the test would still pass when we call `.lockDisplay()` on the question.

The test case above this one is exactly for that particular case.

> This emphasizes the condition "when display was not locked"

I wanted to cover the case "regardless of whether display is locked", but since this requires 2 different setups, I split it into 2 test cases.

WiNloSt commented on e2e/test/scenarios/visualizations/reproductions/30057-30058-map-crash.cy.spec.js:
```diff
@@ -0,0 +1,142 @@
+  });
+});
+
+const addAverageOfLongitudeAggregation = () => {
```
Looks like there are many local functions performing different actions in this test. Have you looked at https://github.com/metabase/metabase/blob/master/e2e/support/helpers/e2e-bi-basics-helpers.js whether existing helpers would help or would some of these new functions should live there?

kamilmielnik replied:
Now it looks much better, thanks!

f957882e95d3a5659c658c6237043e8f730f6251 + 7aa026bcb85f6deebbb2ec271629bae0d43c1125

WiNloSt reviewed (commented):
Left a few comments, but overall this is looking great.

kamilmielnik committed:
Merge branch 'master' into 30057-aggregations-breakouts-fallback

kamilmielnik commented on frontend/test/metabase-lib/lib/Question.unit.spec.js:
```diff
@@ -459,34 +458,67 @@ describe("Question", () => {
+          .setDisplay("funnel")
+          .maybeResetDisplay(sensibleDisplays, sensibleDisplays);
+
+        expect(question.display()).not.toBe("funnel");
+        expect(question.display()).toBe("table");
```
> is there a reason you have 2 assertions like this?

I add these assertions mostly for informative/commentary purposes. This is an explicit way of saying how something used to work in the past, but does not anymore. The small difference in how it works is that this assertion is executed before the one that _catches them all_ - this can potentially give a valuable suggestion for debugging and allows for incremental bug fixing.

kamilmielnik committed:
Add another test case

kamilmielnik commented on frontend/test/metabase-lib/lib/Question.unit.spec.js:
```diff
@@ -459,34 +458,67 @@ describe("Question", () => {
-    describe("maybeUnlockDisplay", () => {
-      it("should keep display locked when it was locked with nonsense display", () => {
+    describe("maybeResetDisplay", () => {
+      it("should do nothing when it was locked with nonsense display", () => {
```
> I think the previous test name was a typo.

I don't think so.

> This makes me think the new test should also be named "should do nothing when it was locked with sensible display"

This is different, but it is a perfectly valid test case :+1: 
Added: e188622ed767601b252ee8edccb636bffe58858f

WiNloSt replied:
You were right. The sensible displays are `table` and `scalar` while the question display is locked with `funnel`.

kamilmielnik committed:
Reuse e2e helpers
- add missing "field" parameter support in addSummaryGroupingField
- add removeSummaryGroupingField

kamilmielnik committed:
Remove redundant condition

kamilmielnik committed:
Update issue id in tests after extracting separate GitHub issue

WiNloSt reviewed (approved):
Looks good to me. 👍

kamilmielnik committed:
Merge branch 'master' into 30057-aggregations-breakouts-fallback

kamilmielnik committed:
Merge branch 'master' into 30057-aggregations-breakouts-fallback
